File: DemoMMOUnity/Assets/Game/Net/Script/Net/CsProtocol/Protocol/Map/PuzzleRewardsData.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using ZJYFrameWork.Collection.Reference;
using ZJYFrameWork.Module.PuzzleNet.Model;
using ZJYFrameWork.Net.CsProtocol.Buffer;
using ZJYFrameWork.Spring.Utils;

namespace ZJYFrameWork.Net.CsProtocol.Protocol.Map
{
    public class PuzzleRewardsData : Model, IPacket, IReference
    {
        /**
 * 奖励的物品id
 */
        public int rewardId { get; set; }

        /**
         * 奖励 type
         */
        public PuzzleRewardsEnum rewardType { get; set; }

        /**
         * 奖励icon
         */
        public string rewardIcon { get; set; }

        /**
         * 奖励 资源
         */
        public string rewardResource { get; set; }

        /**
         * 奖励数量
         */
        public int num { get; set; }

        public short ProtocolId()
        {
            return 203;
        }

        public void Clear()
        {
            rewardId = 0;
            rewardType = 0;
            rewardIcon = string.Empty;
            rewardResource = string.Empty;
            num = 0;
        }

        public static PuzzleRewardsData ValueOf()
        {
            var packet = ReferenceCache.Acquire<PuzzleRewardsData>();
            packet.Clear();
            return packet;
        }
    }

    public class PuzzleRewardsDataRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 203;
        }

        public void Write(ByteBuffer buffer, IPacket packet)
        {
            throw new NotImplementedException();
        }

        public IPacket Read(ByteBuffer buffer)
        {
            var json = StringUtils.BytesToString(buffer.ToBytes());
            var packet = PuzzleRewardsData.ValueOf();
            packet.Clear();
            if (string.IsNullOrEmpty(json))
            {
                return null;
            }

            //解析
            var dict = JsonConvert.DeserializeObject<Dictionary<object, object>>(json);
            foreach (var (key, value) in dict)
            {
                var keyStr = key.ToString();
                var valueStr = value.ToString();
                switch (keyStr)
                {
                    case "rewardId":
                    {
                        packet.rewardId = int.Parse(valueStr);
                    }
                        break;
                    case "rewardType":
                    {
                        var typeNum = int.Parse(valueStr);
                        packet.rewardType = (PuzzleRewardsEnum)typeNum;
                    }
                        break;
                    case "rewardIcon":
                    {
                        packet.rewardIcon = valueStr;
                    }
                        break;
                    case "rewardResource":
                    {
                        packet.rewardResource = valueStr;
                    }
                        break;
                }
            }

            return packet;
        }
    }
}

File: UnityMMOClient/Lua/application/app/net/webSocket/luaProtocol/puzzle/puzzleRewardsData.lua
---@class PuzzleRewardsData
local PuzzleRewardsData = class("PuzzleRewardsData")
function PuzzleRewardsData:ctor()
    --- 奖励数量
    ---@type number
    self.num = 0
    --- 奖励icon
    ---@type string
    self.rewardIcon = string.empty
    --- 奖励的物品id
    ---@type number
    self.rewardId = 0
    --- 奖励资源
    ---@type string
    self.rewardResource = string.empty
    --- 奖励type
    ---@type number
    self.rewardType = 0
end

---@param num number 奖励数量
---@param rewardIcon string 奖励icon
---@param rewardId number 奖励的物品id
---@param rewardResource string 奖励资源
---@param rewardType number 奖励type
---@return PuzzleRewardsData
function PuzzleRewardsData:new(num, rewardIcon, rewardId, rewardResource, rewardType)
    self.num = num --- int
    self.rewardIcon = rewardIcon --- java.lang.String
    self.rewardId = rewardId --- int
    self.rewardResource = rewardResource --- java.lang.String
    self.rewardType = rewardType --- int
    return self
end

---@return number
function PuzzleRewardsData:protocolId()
    return 203
end

---@return string
function PuzzleRewardsData:write()
    local message = {
        protocolId = self:protocolId(),
        packet = {
            num = self.num,
            rewardIcon = self.rewardIcon,
            rewardId = self.rewardId,
            rewardResource = self.rewardResource,
            rewardType = self.rewardType
        }
    }
    local jsonStr = JSON.encode(message)
    return jsonStr
end

---@return PuzzleRewardsData
function PuzzleRewardsData:read(data)

    local packet = self:new(
            data.num,
            data.rewardIcon,
            data.rewardId,
            data.rewardResource,
            data.rewardType)
    return packet
end

--- 奖励数量
---@return number 奖励数量
function PuzzleRewardsData:getNum()
    return self.num
end
--- 奖励icon
---@type  string 奖励icon
function PuzzleRewardsData:getRewardIcon()
    return self.rewardIcon
end

--- 奖励的物品id
---@return number 奖励的物品id
function PuzzleRewardsData:getRewardId()
    return self.rewardId
end
--- 奖励资源
---@type  string 奖励资源
function PuzzleRewardsData:getRewardResource()
    return self.rewardResource
end

--- 奖励type
---@return number 奖励type
function PuzzleRewardsData:getRewardType()
    return self.rewardType
end


return PuzzleRewardsData
